File: Python/Make.py
import numpy as np
from FCM_Function import fcm_function as FCM
from scipy.io import savemat
import pandas as pd

def Make(train_data):
    train_data = pd.read_csv(train_data)
    data_num, attribute_num = train_data.shape
    data = train_data[:, :-1]  # All columns except the last one (features)
    nhan1 = train_data[:, -1]  # Last column (labels)

    cluster = np.zeros(attribute_num)  # Initialize cluster type for each feature

    # Determine the number of clusters based on unique values in each feature
    for i in range(attribute_num):
        temp = len(np.unique(train_data[:, i]))
        if temp == 2:
            cluster[i] = 2
        elif temp == 3:
            cluster[i] = 3
        else:
            cluster[i] = 3

    m = 2
    esp = 0.01
    maxTest = 200
    center_vector = [None] * attribute_num
    centers = [None] * (attribute_num - 1)

    X1 = np.zeros((data_num, attribute_num))

    # Perform Fuzzy C-Means Clustering for each feature
    for feature_index in range(attribute_num):
        feature_data = train_data[:, feature_index]
        V = np.zeros((3, 1))  # Initialize cluster centers

        min_value = np.min(train_data[:, feature_index])
        max_value = np.max(train_data[:, feature_index])
        delta = max_value - min_value
        
        if cluster[feature_index] == 2:
            V[0, 0] = min_value - 0.5
            V[1, 0] = max_value
        else:
            V[0, 0] = min_value
            V[1, 0] = min_value + delta / 2
            V[2, 0] = max_value

        # Perform Fuzzy C-Means Clustering
        fcm = FCM(n_clusters=int(cluster[feature_index]))
        fcm.fit(feature_data.reshape(-1, 1))  # Reshape to 2D array for FCM
        U = fcm.u.T  # Transpose membership matrix
        center = fcm.centers

        center_vector[feature_index] = center[:, 0]

        # Assign clusters based on the maximum membership value
        for i in range(data_num):
            maximum = np.max(U[i, :])
            for j in range(int(cluster[feature_index])):
                if maximum == U[i, j]:
                    X1[i, feature_index] = j + 1  # +1 to match 1-based indexing

    # Label calculation (same as in the original MATLAB code)
    label = train_data[:, 0] + 1  # Labels should start from 1

    # Prepare the ruleList1 matrix
    ruleList1 = X1.astype(int)

    # Calculate the A matrix (relationship matrix)
    r1, r2 = ruleList1.shape
    A = np.zeros((r1, int((r2 - 3) * (r2 - 2) * (r2 - 1) / 6)))  # A matrix size calculation
    j = 0
    for i in range(r1):
        for l in range(r2 - 3):
            for h in range(l + 1, r2 - 2):
                for s in range(h + 1, r2 - 1):
                    count = 0
                    for t in range(r1):
                        if (ruleList1[t, l] == ruleList1[i, l] and
                                ruleList1[t, h] == ruleList1[i, h] and
                                ruleList1[t, s] == ruleList1[i, s]):
                            count += 1
                    A[i, j] = count / r1
                    j += 1

    # Calculate the B matrix (another relationship matrix)
    B = np.zeros_like(A)
    T = np.sum(A, axis=1)
    for i in range(r1):
        j = 0
        for l in range(r2 - 2):
            for h in range(l + 1, r2 - 1):
                dem1 = 0
                dem2 = 0
                for t in range(r1):
                    if (ruleList1[t, l] == ruleList1[i, l] and ruleList1[t, r2 - 1] == ruleList1[i, r2 - 1]):
                        dem1 += 1
                    if (ruleList1[t, h] == ruleList1[i, h] and ruleList1[t, r2 - 1] == ruleList1[i, r2 - 1]):
                        dem2 += 1
                M1 = dem1 / r1
                M2 = dem2 / r1
                B[i, j] = T[i] * min(M1, M2)
                j += 1

    # Save results to a .mat file
    result_dict = {
        'X1': X1,
        'label': label,
        'A': A,
        'B': B,
        'ruleList1': ruleList1
    }

    savemat('output.mat', result_dict)

    return X1, A, B, center_vector

# Example usage:
# train_data is assumed to be a numpy array or a pandas DataFrame with your training data.
  # Example random data
X1, A, B, center_vector = Make('Tan1.txt')

File: Matlab/Make.m
function [X1 A B cluster center_vector] = Make(train_data)

[data_num attribute_num] = size(train_data);
data=train_data(:,1:attribute_num-1);
nhan1=train_data(:,attribute_num);

cluster = zeros(1, attribute_num);                  
                                     
for i=1:attribute_num
    temp = length(unique(train_data(:,i)));
    if temp == 2
        cluster(i) = 2;
    elseif temp == 3
        cluster(i) = 3;
    else
        cluster(i) = 3;
    end
end

m = 2;
esp = 0.01;
maxTest = 200;
center_vector = cell (1, attribute_num);
centers = cell(attribute_num-1, 1);
for feature_index=1:attribute_num
    feature_data = train_data(:,feature_index);
    V = 0;
    min_value = min(train_data(:,feature_index));
    max_value = max(train_data(:,feature_index));
    delta = max_value - min_value;
    if cluster(feature_index) == 2
        V(1,1) = min_value - 0.5;
        V(2,1) = max_value;

%     elseif cluster(feature_index) == 3
    else
        V(1,1) = min_value;
        V(2,1) = min_value + delta/2;
        V(3,1) = max_value;
  
    end
    [center,U] = FCM_Function(feature_data,cluster(feature_index),V,m,esp,maxTest);
    U = U';
    center_vector{feature_index}(:,1) = center(:,1);
    for i=1:data_num
        maximum = max(U(i,:));
        for j=1:cluster(feature_index)
            if (maximum == U(i,j))
                X1(i,feature_index) = j;
            end
        end
    end
    
end
% X1=[X1 nhan1];

label = train_data(:,1)+1;
label=label;
ruleList1=X1;
% T1=0;
% T2=0;
% for i=1:data_num
%     if ruleList1(i,attribute_num)==1
%         if label(i)==1
%             T1=T1+1;
%         else
%             T2=T2+1;
%         end
%     end
% end
% ruleList1(:,attribute_num);
% if T1<T2
%     for i=1:data_num
%         if ruleList1(i,attribute_num)==1
%             ruleList1(i,attribute_num)=2;
%         else
%             ruleList1(i,attribute_num)=1;
%         end
%     end
% end



% % -------------------
% 
ruleList1=[1 1 1 1 1 1;2 3 2 2 2 2;1 3 1 1 1 1;1 1 1 2 1 1;2 3 2 3 2 2;2 3 3 3 3 2];
% 
[r1 r2]= size(ruleList1);
%Sinh do thi 
%Tinh A
for i =1:r1
    j=1;
    for l=1:r2-3
        for h=l+1:r2-2
            for s=h+1:r2-1
                dem=0;
                for t=1:r1
                    if (ruleList1(t,l)==ruleList1(i,l))&(ruleList1(t,h)==ruleList1(i,h))&(ruleList1(t,s)==ruleList1(i,s))
                            dem=dem+1;
                    end
                end
                
            A(i,j)=dem/r1;
            j=j+1;
            end
        end
    end
end
ruleList1;
A
% C=sum(A');
%Tinh tong A theo nhãn lu?t
[k1 k2]= size(A);
T=sum(A')
% for i=1:2
%     for j=1:k2
%         tong=0;
%         for l=1:k1
%             if ruleList1(l,r2)==i
%                 tong=tong+A(l,j);
%             end
%         end
%         T(i,j)=tong;
%     end
% end
% T
%tìm B
r1;
for i=1: r1
    j=1;
    for l=1:r2-2
        for h=l+1:r2-1
%             for s=h+1:r2-1
                dem1=0;
                dem2=0;
%                 dem3=0;
                for t=1:r1
                    if (ruleList1(t,l)==ruleList1(i,l))&(ruleList1(t,r2)==ruleList1(i,r2))
                        dem1=dem1+1;
                    end
                    if (ruleList1(t,h)==ruleList1(i,h))&(ruleList1(t,r2)==ruleList1(i,r2))
                        dem2=dem2+1;
                    end
%                     if (ruleList1(t,s)==ruleList1(i,s))&(ruleList1(t,r2)==ruleList1(i,r2))
%                         dem3=dem3+1;
%                     end
                end
                M1=dem1/r1;
                M2=dem2/r1;
%                 M3=dem3/r1;
              
%                 for l1=1:2
%                     if ruleList1(i,r2)==l1
                B(i,j)=T(i)*min(M1,M2);
                j=j+1;
%             end
        end
    end
end
B
   

% nhan=[1 2];
% n1=size(nhan)
% for i=1:r1
%     for j=1:r2
%         dem=0;
%         for l=1:r1
%             if (ruleList1(i,r2)==ruleList1(l,r2))&(ruleList1(i,j)==ruleList1(l,j))
%                 dem=dem+1;
%             end
%         end
%         
%         B(i,j)=C(i)*dem/r1;
%     end
% end
size(A);
size(B);
% 
% 
% 
% 
% filename = strrep(train_data,'.txt','.mat');
% filename = strrep(filename, 'Database', 'FIS');
% addpath('..\output');
% save(['..\output\RuleList.mat'], 'ruleList');
% save(['..\output\RuleList.mat'], 'ruleList_var','-append');
% save(['..\output\FIS_Breast.mat'], 'sigma_M');
% save(['..\output\FIS_Breast.mat'], 'sigma_M_var', '-append');

